dcrec/secp256k1/ecdsa/signature_test.go

```diff
@@ -271,6 +274,464 @@ func TestSignatureSerialize(t *testing.T) {
+				"private key: %x\npublic key: %x", sig.Serialize(), hash,
+				privKey.Serialize(), pubKey.SerializeCompressed())
+		}
+
+		// Change a random bit in the signature and ensure the bad signature
+		// fails to verify the original message.
+		badSig := *sig
+		randByte := rng.Intn(32)
+		randBit := rng.Intn(7)
+		if randComponent := rng.Intn(1); randComponent == 0 {
+			badSigBytes := badSig.r.Bytes()
+			badSigBytes[randByte] ^= 1 << randBit
+			badSig.r.SetBytes(&badSigBytes)
+		} else {
+			badSigBytes := badSig.s.Bytes()
+			badSigBytes[randByte] ^= 1 << randBit
+			badSig.s.SetBytes(&badSigBytes)
+		}
+		if badSig.Verify(hash[:], pubKey) {
+			t.Fatalf("verified bad signature\nsig: %x\nhash: %x\n"+
+				"private key: %x\npublic key: %x", badSig.Serialize(), hash,
+				privKey.Serialize(), pubKey.SerializeCompressed())
+		}
+
+		// Change a random bit in the hash that was originally signed and ensure
+		// the original good signature fails to verify the new bad message.
+		badHash := make([]byte, len(hash))
+		copy(badHash, hash[:])
+		randByte = rng.Intn(len(badHash))
+		randBit = rng.Intn(7)
+		badHash[randByte] ^= 1 << randBit
+		if sig.Verify(badHash[:], pubKey) {
+			t.Fatalf("verified signature for bad hash\nsig: %x\nhash: %x\n"+
+				"pubkey: %x", sig.Serialize(), badHash,
+				pubKey.SerializeCompressed())
+		}
+	}
+}
+
+// TestSignFailures ensures the internal ECDSA signing function returns an
+// unsuccessful result when particular combinations of values are unable to
+// produce a valid signature.
+func TestSignFailures(t *testing.T) {
+	tests := []struct {
+		name  string // test description
+		key   string // hex encoded private key
+		hash  string // hex encoded hash of the message to sign
+		nonce string // hex encoded nonce to use in the signature calculation
+	}{{
+		name:  "zero R is invalid (forced by using zero nonce)",
+		key:   "0000000000000000000000000000000000000000000000000000000000000001",
+		hash:  "c301ba9de5d6053caad9f5eb46523f007702add2c62fa39de03146a36b8026b7",
+		nonce: "0000000000000000000000000000000000000000000000000000000000000000",
+	}, {
+		name:  "zero S is invalid (forced by key/hash/nonce choice)",
+		key:   "0000000000000000000000000000000000000000000000000000000000000001",
+		hash:  "393bec84f1a04037751c0d6c2817f37953eaa204ac0898de7adb038c33a20438",
+		nonce: "4154324ecd4158938f1df8b5b659aeb639c7fbc36005934096e514af7d64bcc2",
+	}}
+
+	for _, test := range tests {
+		privKey := hexToModNScalar(test.key)
+		hash := hexToBytes(test.hash)
+		nonce := hexToModNScalar(test.nonce)
+
+		// Ensure the signing is NOT successful.
+		sig, _, success := sign(privKey, nonce, hash[:])
+		if success {
+			t.Errorf("%s: unexpected success -- got sig %x", test.name,
+				sig.Serialize())
+			continue
+		}
+	}
+}
+
+// TestVerifyFailures ensures the ECDSA verification function returns an
+// unsuccessful result for edge conditions.
+func TestVerifyFailures(t *testing.T) {
```

Comment:
It doesn't make a big difference since there are not a huge amount of tests, but it might be worth running the tests in parallel as a standard since it will eventually add up across all the various test suites:
```suggestion
func TestVerifyFailures(t *testing.T) {
    t.Parallel()
```

Reply:
Agreed.  I added it to all of the new tests introduced as well.